### PR TITLE
sys: sema: semaphore bugfixes

### PR DESCRIPTION
**Pull request by lebrush**
While testing the semaphore implementation in the real world (not only existing tests) I found two issues introduced by #6155.

 - ~~Timeout 0 was expected to lock indefinitely (0 for no timeout meant "waits forever").~~ we'll keep it like this and update lwip to call the blocking method instead
 - Value has to be re-tested after owning the mutex as otherwise it can't be guaranteed that the value didn't change in the meantime.

### CODE REVIEW

**Comment by kaspar030** [comment]
> Timeout 0 was expected to lock indefinitely (0 for no timeout meant "waits forever").

Personally I find "wait zero time" for "don't block" more intuitive...
We had that discussion for sock_udp, and should IMHO use the same semantics.

**Comment by lebrush** [comment]
This is how I understood it when I re-implemented it as well... but the previous behavior was not like this. I've a mixed feeling about it because sometimes times `timeout=0` (or -1) means wait forever. But if the current approach in the `sock_upd` is not to wait, let's keep it the same :+1: 

I'll quickly update and test it.

**Comment by lebrush** [comment]
@kaspar030 updated as agreed.

File: sys/sema/sema.c [hunk 1]
```diff
@@ -44,53 +44,54 @@ void sema_destroy(sema_t *sema)
...
     assert(sema != NULL);
 
     if (sema->state != SEMA_OK) {
         return -ECANCELED;
     }
 
+    int ret = (block) ? -ETIMEDOUT : -EAGAIN;
     unsigned old = irq_disable();
-    if (sema->value == 0) {
+    while (sema->value == 0 && block) {
```
Comment (kaspar030) [inline sys/sema/sema.c]: please put ```sema->value == 0```in "()"

Comment (lebrush) [inline sys/sema/sema.c]: done

**Comment by miri64** [comment]
> We had that discussion for sock_udp, and should IMHO use the same semantics.

lwIP's and POSIX semaphores probably need adaptation then.

**Comment by lebrush** [comment]
lwip is adapted in this PR. Posix doesn't need adaption since there's no
wait with timeout there :-)

**Comment by miri64** [comment]
@kaspar030 happy?

**Comment by lebrush** [comment]
@kaspar030 ping

**Comment by brummer-simon** [comment]
I tested the PR and finally the semaphore is use in a private project works as expected 👍

**Pull request merged**